Major refactor of cover platform and config flow support tweaks for cover platform

andrewmeepos opened the pull request:
This is a major rework and refactor of cover.py.
**Depends on https://github.com/rospogrigio/localtuya-homeassistant/pull/6**
**Should be merged after #8 for the cleanest merging.**

See https://github.com/postlund/localtuya-homeassistant/pull/1 for previous conversation with @postlund on this PR.

A few things to note about my pushed code:

1. This is a a pretty major rework of cover.py. I refactored almost all of the functions to align with Home Assistant best practices for devices generally and for covers specifically, per https://developers.home-assistant.io/docs/core/entity/cover/ and https://developers.home-assistant.io/docs/asyncio_working_with_async/.

2. None of my "set" functions (open, close, set position, etc.) in cover.py call anything in __init__.py other than set_value. As rospogrigio mentioned, and with which I completely agree, pytuya should be a minimal library to communicate with Tuya devices. Handling of specific device characteristics, attributes, functions, etc., should be done on localtuya's end.

3. Here's my lovelace code which shows the blind using a few different card types (some stock, some custom).
```
- entities:
          - buttons_position: left
            entity: cover.test1007
            invert_percentage: true
            name: Test Blind
            title_position: bottom
        title: Test Blind
        type: 'custom:shutter-card'
      - entities:
          - entity: cover.test1007
            icon: 'mdi:blinds'
            name: Left Blind
            type: 'custom:slider-entity-row'
          - type: divider
          - entity: cover.test1007
            icon: 'mdi:blinds'
            name: Left Blind
        full_row: true
        show_header_toggle: false
        title: Test Blind
        type: entities
```

And that gets you:
<img width="390" alt="Screen Shot 2020-09-11 at 10 10 51 AM" src="https://user-images.githubusercontent.com/28188533/92954645-a2721780-f418-11ea-98c3-441ed0b28b1f.png">

(Note: The reason the top card shows 81% and the bottom card shows 80% is because the bottom card rounds to the nearest 10%).

4. There are a couple FIXME/TODO:

- I still want to think about whether it's possible during config_flow to only allow the user to select from the dps list for the following characteristics: GET_POSITION_KEY, SET_POSITION_KEY, COVER_COMMAND_KEY, LAST_MOVEMENT_KEY. Entering an integer with no list to choose from is not terrible, but it's not optimal. As a stopgap measure, perhaps we print the dps data in the config_flow setup screen so that people can refer to it:
<img width="421" alt="Screen Shot 2020-09-11 at 10 08 08 AM" src="https://user-images.githubusercontent.com/28188533/92954875-0f85ad00-f419-11ea-8952-33fb0c34bc53.png">

- I need to add code to validate that OPEN_CMD, CLOSE_CMD, and STOP_CMD are all in lowercase ("open" vs "OPEN" or "Open"--only works when all lowercase). Or, alternatively, let people use whatever they want in config and then parse it as lowercase in cover.py. Or both. Easy fix but not a big deal and will do it later.

- When you hit one button on the lovelace UI (such as open) and then VERY QUICKLY (within less than a second or so) hit another button (such as stop), the second command sometimes does not register. This might just be a limitation of Tuya devices--they don't like getting inundated with rapid connections.

- When you open or close the blind, and then hit stop, the current_cover_position immediately updates to the correct value. However, when you set a specific position, and the cover opens/closes to that position, the current_cover_position does not update until the next update call after the opening or closing is complete, which can be as long as 30 seconds, but is usually closer to 15 seconds.

rospogrigio commented:
Regarding keeping track of opening percentage, I had thought a lot about it and I think it's not a good idea, or at least I would make it optional. This because if you have a cover (like mine) that has also physical buttons (which I believe is the most common situation) there is no way for the device to know if a manual opening or closing has occurred, and for how long. The positioning info is unreliable for these devices, as a consequence.

andrewmeepos commented:
> Regarding keeping track of opening percentage, I had thought a lot about it and I think it's not a good idea, or at least I would make it optional. This because if you have a cover (like mine) that has also physical buttons (which I believe is the most common situation) there is no way for the device to know if a manual opening or closing has occurred, and for how long. The positioning info is unreliable for these devices, as a consequence.

Hi @rospogrigio, thanks for the note. I'm curious to see how your device compares to mine. This will help us think of the best way to approach this situation. Can you run the following two scenarios for me:

**Scenario A**
1. Using the physical buttons, start closing your cover device (by close, I mean cover up your window with the cover)
2. Wait until the cover is completely covering your window and the cover movement has stopped 
2. Run `tuya-cli get --id YOURID --key YOURKEY --ip YOURID --protocol-version YOURPROTOCOL --all`
3. Output your results below:



4. Using the physical buttons, start opening your cover device
5. After 3 seconds, use the physical button to stop the opening
6. Run `tuya-cli get --id YOURID --key YOURKEY --ip YOURID --protocol-version YOURPROTOCOL --all`
7. Output your results below:



**Scenario B**
1. Using Home Assistant, start closing your cover device (by close, I mean cover up your window with the cover)
2. Wait until the cover is completely covering your window and the cover movement has stopped 
2. Run `tuya-cli get --id YOURID --key YOURKEY --ip YOURID --protocol-version YOURPROTOCOL --all`
3. Output your results below:



4. Using Home Assistant start opening your cover device
5. After 3 seconds, use Home Assistant to stop the opening of the cover device
6. Run `tuya-cli get --id YOURID --key YOURKEY --ip YOURID --protocol-version YOURPROTOCOL --all`
7. Output your results below:

rospogrigio commented:
OK I'll try it but first I have to fix tuya-cli because it has never worked with my cover devices, but now I should know how to fix it.
Will keep you posted.

gd71 commented:
Hello, I'm using 3 cover switches with the current implementation of local tuya.
[(https://www.amazon.fr/gp/product/B07PNV9XKQ/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1)]
on any scenario described above, the result of the command is:
`{ devId: '27751547fcf5c48598b8',
  dps: { '1': 'stop', '9': 15, '101': true } }
`
So I guess the remark from @rospogrigio is relevant.
dps 1 is for the well known commands ('on','off','stop' in my case), 9 is for the trip time (in seconds) I configured via the smart app then 101 looks to me for the backlight.
Not very skilled on debugging/developing on such kind devices, but at least I'm at your disposal for some tests.
Keep up your great job!

andrewmeepos commented:
@rospogrigio, I forgot that tuya-cli does not work with your device. Can you get me the output of those 4 scenarios above (A.3, A.7, B.3, B.7) using a different method. Instead of using tuya-cli, let's use localtuya's debug output. In `pytuya/__init__.py`, ensure that you have the the two `log.debug('decrypted result=%r', result)` lines in your file and not commented out.

```
elif result.startswith(PROTOCOL_VERSION_BYTES_31):
            # got an encrypted payload, happens occasionally
            # expect resulting json to look similar to:: {"devId":"ID","dps":{"1":true,"2":0},"t":EPOCH_SECS,"s":3_DIGIT_NUM}
            # NOTE dps.2 may or may not be present
            result = result[len(PROTOCOL_VERSION_BYTES_31):]  # remove version header
            result = result[16:]  # remove (what I'm guessing, but not confirmed is) 16-bytes of MD5 hexdigest of payload
            cipher = AESCipher(self.local_key)
            result = cipher.decrypt(result)
            log.debug('decrypted result=%r', result)
            if not isinstance(result, str):
                result = result.decode()
            result = json.loads(result)
        elif self.version == 3.3: 
            cipher = AESCipher(self.local_key)
            result = cipher.decrypt(result, False)
            log.debug('decrypted result=%r', result)
```

Then, enable debug logging in `configuration.yaml`:
```
#Logging
logger:
  default: info
  logs:
    custom_components.localtuya: debug
```

Now, run the 4 scenarios above (A.3, A.7, B.3, B.7), and after each one, wait until the next `decrypted_result` posts to the Home Assistant log, and then grab that output.

andrewmeepos commented:
> Hello, I'm using 3 cover switches with the current implementation of local tuya.
> (https://www.amazon.fr/gp/product/B07PNV9XKQ/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1)
> on any scenario described above, the result of the command is:
> `{ devId: '27751547fcf5c48598b8', dps: { '1': 'stop', '9': 15, '101': true } } `
> So I guess the remark from @rospogrigio is relevant.
> dps 1 is for the well known commands ('on','off','stop' in my case), 9 is for the trip time (15 seconds here) I configured via the smart app then 101 looks to me for the backlight.
> Not very skilled on debugging/developing on such kind devices, but at least I'm at your disposal for some tests.
> Keep up your great job!

Thank you for this, it's very helpful. As you said, your cover device opens/closes based on a set amount of time. It does not report current position, nor let you set current position. It's entirely time based.

That's very different from my device, which does not use timers. Rather, it reports the current position ('3') and lets me set the current position directly ('2').
```
  dps: {
    '1': 'open',
    '2': 51,
    '3': 51,
    '5': false,
    '7': 'closing',
    '8': 'cancel',
    '9': 0,
    '10': 0,
    '11': 38610
  }
```

I think that @rospogrigio's device is similar to yours, as his commands are also `on` `off` and `stop` (whereas mine are `open` `close` and `stop`).

However, I want to wait until @rospogrigio provides his output for those 4 scenarios just to be certain.

If @rospogrigio's output looks similar to yours, then I think a potential solution is as follows:
1. Add another configuration option to device setup called `Operation Mode` with the available options as `Timed` and `Position`.
2. If you set `Position`, then everything operates exactly as this PR currently stands.

3. If you set `Timed`, then you don't need to enter GET_POSITION_KEY, SET_POSITION_KEY in the device configuration (or you can enter them, but localtuya will simply ignore them). On the back-end, set position is disabled. Open, close, stop work as advertised. 

@rospogrigio I know you used to have some code for setting position of time-based cover devices:
```
def set_cover_position(self, **kwargs):
        """Move the cover to a specific position."""
        newpos = float(kwargs["position"])
        currpos = self.current_cover_position
        posdiff = abs(newpos - currpos)
#       25 sec corrisponde alla chiusura/apertura completa
        mydelay = posdiff / 2.0
        if newpos > currpos:
            self.open_cover()
        else:
            self.close_cover()
        sleep( mydelay )
        self.stop_cover()
        self._position = 50  # newpos
```
Can you explain how this code worked in conjunction with your device. It's confusing to me, because you're setting the `self._position=50` after each call of this function, so obviously that is a proxy for device position since you can't get it directly from the device. However, when you run `set_cover_position`, you would always be comparing the new position to `50`, right? I'm not quite sure how this works. What if the cover's actual position is currently `10`, and you use this code to `set_cover_position` to `50`? It looks to me like nothing would happen, since `newpos` and `currpos` would both be `50`, `posdiff` would be zero, `self.close_cover()` would be called but `mydelay` would be `0` so the function would then immediately call `self.stop_cover()`.

andrewmeepos commented:
My general thoughts are that if a cover device doesn't technically support setting the position, then we probably shouldn't include a hacked together attempt to set the position.

And on that train of thought, if a device does not support setting the position, we should ensure that `SUPPORT_SET_POSITION` is not sent to Home Assistant for the device. That might mean that we need to dynamically set the device's supported features based on whether the user sets the new `Operation Mode` in the initial device config to `Timed` or `Position`.

Those are my thoughts on the matter. @postlund, curious to hear your thoughts (on how we should approach cover devices that don't technically report the current position or let you set the new position).

postlund commented:
I tend to agree with you @andrewmeepos. If we can't provide a reliable position it's better if we don't. The user will know this. Since Home Assistant provides a template for covers, users can toll their own solution:

https://www.home-assistant.io/integrations/cover.template/#position_template

postlund commented:
And yeah, we must not return `SUPPORT_SET_POSITION` in case we don't support it. But that should be rather simple if we depend on an optional DPS value (no value = not supported).

gd71 commented:
please note that the cover device is managing the time by itself. Could be a workaround to finally set position to 50 in order to let the widget always proposing close and open actions as HASS does not know the real position.
Fully shared, if the device is not in a situation to feedback the position then don't try to guess it, there will be more issues than benefits

rospogrigio commented:
@andrewmeepos , I will try to get the output you need as soon as I can.
Regarding the code for positioning, I knew this question would come, sooner or later... Well, it's a kind of a hack actually. The sense is to allow a partial opening or closing of the desired amount, regardless of the current position, which shall be known a priori by the user. It just doubles the posDiff and sends it as a positioning command. To better explain, the cover believes it is always at 50, and you can give positioning commands like 100 (fully open) or 0 (fully close). Setting 75 would open half way, 25 would close it half way (the 25 difference from 50 is doubled). Obviously this movement is effective only if the user is aware of how open the cover is. I just added it because if I wanted, let's say, to open it just a 10% more than it is now, I just needed to set 55, instead of issuing an open+stop commands. 
It is not of vital importance, if you want to get rid of this little hack. Hope it's clearer now, though.
Bye!

rospogrigio commented:
Here is the output:
DEBUG:localtuya.pytuya:decrypted result='{"dps":{"1":"stop","101":false},"t":1599985046}'

I only get the status and the 101, which is the backlight status. I"d leave the behavior configurable, with options Position/Hacked position/None.
Your thoughts?

gd71 commented:
@rospogrigio , I wonder whether the hacked position could be achieved through the template cover, https://www.home-assistant.io/integrations/cover.template/#position_template as suggested by @postlund and script for setting position. But I'm very new to HA and not yet comfortable with these features

andrewmeepos commented:
`I'd leave the behavior configurable, with options Position/Hacked position/None.`

I could be ok with including "Hacked Position" as an option, as long as we **very clearly** explain what it is that "Hacked Position" does. It is definitely not proper Home Assistant behavior.

I have no experience with `position_template`, so I won't comment on that.

gd71 commented:
In my case, the trip-time is at dps 9 and most probably can be set by HA. I guess this can be detected automatically.
If you want to implement the "hacked position", my guess is that it is up to the user to set the trip-time. My current tests on current implementation is not accurate for me as it looks like 25s is hardcoded. Also, as you mention, we must be sure that HA continue working when sleep instruction is ongoing

andrewmeepos commented:
I would like to propose that we proceed with merging this pull request.

After that, @rospogrigio, you can open a new pull request which adds the functionality to use the "Hacked Position" open, and we can continue the discussion there.

rospogrigio commented:
@andrewmeepos , I would like to merge #11 and #12 before getting into this. Can you try to merge from #11 and fix the conflicts from there? If you still have problems I can help but give me some time.

andrewmeepos commented:
> @andrewmeepos , I would like to merge #11 and #12 before getting into this. Can you try to merge from #11 and fix the conflicts from there? If you still have problems I can help but give me some time.

Yes, i'll start with #11 and build this back on top of it.

andrewmeepos commented:
Refer to https://github.com/rospogrigio/localtuya-homeassistant/pull/19

postlund commented:
You can always just "push -f" to your branch to start over. There's really no need to close the PR.

andrewmeepos commented:
@postlund Ok, finally figured out how to rebase and force push. Anyways, the one error holding me up on this is below. I get the error every time I finish setting up a cover device from the config flow.

```
2020-09-16 01:17:29 ERROR (MainThread) [homeassistant.components.cover] Error while setting up localtuya platform for cover
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 184, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/localtuya/cover.py", line 111, in async_setup_entry
    TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),
KeyError: 'friendly_name'`
```

rospogrigio commented:
As I wrote in #19, I'm getting the same error so that's not your fault. Will try to fix it when I get the time.

andrewmeepos commented:
Do you get this error when setting up switches too, or just covers?

andrewmeepos commented:
Once we get the friendly_name error fixed, this should be good to merge.

postlund commented:
@andrewmeepos You can try #20 and see if it works.

andrewmeepos commented:
I believe that this is ready to merge. @rospogrigio or @postlund will you review the changes?

rospogrigio commented:
I reviewed it and added comments, also please revert #20 since it's being discussed. Moreover, my cover doesn't work with this PR even though I don't get errors in the logs, need to figure out what's going on but I'll have a super busy day so I won't be able to have a look at it today, sorry.

andrewmeepos commented:
> I reviewed it and added comments, also please revert #20 since it's being discussed. Moreover, my cover doesn't work with this PR even though I don't get errors in the logs, need to figure out what's going on but I'll have a super busy day so I won't be able to have a look at it today, sorry.

Reverting #20 breaks setup, hence why I implemented #20 in the first place. It's this line here:
`                TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),`
That throw a key_error
`KeyError: 'friendly_name'`

postlund commented:
> Reverting #20 breaks setup, hence why I implemented #20 in the first place. It's this line here:
> 
> `                TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),`
> 
> and specifically this portion of it
> 
> `config_entry.data[CONF_FRIENDLY_NAME]`
> 
> That throws the below error and causes config flow setup to fail (I haven't tested yaml setup):
> 
> `KeyError: 'friendly_name'`

That error should be solved in a separate PR to not make a mess. So it should not be included here, you can just say that you depend on another PR instead and then we wait until that is merged before merging this.

andrewmeepos commented:
@postlund I fixed the key error. All I had to do was change this line in cover.py:
`                TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),`
to
`                TuyaCache(device, device_config[CONF_FRIENDLY_NAME]),`

postlund commented:
> @postlund I fixed the key error. All I had to do was ...
> I'm not sure if this is the correct way to go about things, but this did fix things on my end.

It is not, we are still working on the details. So just remove these changes and wait for it to be fixed in another PR.

andrewmeepos commented:
> > @postlund I fixed the key error. All I had to do was ...
> > I'm not sure if this is the correct way to go about things, but this did fix things on my end.
> 
> It is not, we are still working on the details. So just remove these changes and wait for it to be fixed in another PR.

Ok. Fully reverted #20 in https://github.com/rospogrigio/localtuya-homeassistant/pull/10/commits/508fd05df7404c3de2a664e7fc1aa5fec01feb2c. 

Can't set up my cover anymore, but i'll just run a modified version until we get this fixed in master.

rospogrigio reviewed (commented):
> Any other changes that need to be made to this PR?

Other than those already added, you should now merge from #25 since a major refactoring has been done, sorry about this... please try to comply with the new architecture.

andrewmeepos commented:
> > Any other changes that need to be made to this PR?
> 
> Other than those already added, you should now merge from #25 since a major refactoring has been done, sorry about this... please try to comply with the new architecture.

Sorry, I missed your comments! I agree about rebasing this to run on top of #25. I'll update this PR later tonight.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -49,33 +54,49 @@
+            vol.Optional(CONF_OPEN_CMD, default=DEFAULT_OPEN_CMD): str,
+            vol.Optional(CONF_CLOSE_CMD, default=DEFAULT_CLOSE_CMD): str,
+            vol.Optional(CONF_STOP_CMD, default=DEFAULT_STOP_CMD): str,
+            vol.Optional(CONF_SET_POSITION_DPS, default=DEFAULT_SET_POSITION_DPS): cv.positive_int,
```
Did you verify that this worked? I've had problems using `cv.string` with config flows, doesn't seem to be supported. Is it ok with `cv.positive_int`? I got an exception.

andrewmeepos replied:
I have had no problem with `str` or `cv.positive_int`.

postlund replied:
Great! 👍

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -97,6 +118,15 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
                 TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),
                 config_entry,
                 device_config[CONF_ID],
```
Remove unnecessary old code.

andrewmeepos replied:
I deleted:
```
TuyaCache(device, config_entry.data[CONF_FRIENDLY_NAME]),
                config_entry,
                device_config[CONF_ID],
```

So that it now reads:
```
    for device_config in entities_to_setup:
        dps[device_config[CONF_ID]] = None
        covers.append(
            LocaltuyaCover(
            )
        )
```

Is that correct?

postlund replied:
The original code was correct, I was referring to the commented code that was used before.

andrewmeepos replied:
Gotcha, I thought my edit seemed wrong. Thanks for the clarification.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -105,6 +135,7 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up of the Tuya cover."""
+    _LOGGER.info("running setup_platform")
```
You can remove this (at least not keep it as `info`). This goes for most of the log points which are called a lot in this file.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -186,19 +220,194 @@ def __init__(
             "Initialized tuya cover  [{}] with switch status [{}] and state [{}]".format(
                 self.name, self._status, self._state
+
+    # def __init__(self, device, friendly_name, coverid, open_cmd, close_cmd, stop_cmd, set_position_dps, get_position_dps, last_movement_dps):
```
Remove old code.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -186,19 +220,194 @@ def __init__(
         )
 
+    @property
+    def device_info(self):
```
You can remove `device_info` altogether as that is handled by `LocalTuyaEntity` now.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -186,19 +220,194 @@ def __init__(
+        }
+
+    @property
+    def open_cmd(self):
```
Most of these properties are not needed and can be removed, they just clutter the code. The only ones we need are those that override something in `CoverEntity`. You can compare with this and remove the rest:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/cover/__init__.py

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -186,19 +220,194 @@ def __init__(
+        #_LOGGER.info("is_closed returning false")
+        return False
+
+    def update(self):
```
You should not override `update` anymore. Rename this to `status_updated`. It will be called when an update has been performed and the "status" is available via `self._status` instead. Everything related to `self._available` can be removed as well, since exceptions are caught and handles availability automatically. You can look here if you want to know how it works:

https://github.com/rospogrigio/localtuya-homeassistant/blob/c266e0f90419aa52f0071a96d3258c022ebe9b51/custom_components/localtuya/__init__.py#L158

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -186,19 +220,194 @@ def __init__(
+            _LOGGER.info("self._cached_status set =%s", self._cached_status)
+            
+            _LOGGER.info("about to call self._update_last_command()")
+            self._update_last_command()
```
I believe that you can just inline the variable assignments of al these calls here. The names of the variables make it quite clear what's going on (at least the methods doesn't add much). If you really need debugging here, add one line that prints every value (compactly) with `debug` level. It gets very chatty otherwise.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -207,87 +416,85 @@ def available(self):
-        supported_features = (
-            SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_STOP | SUPPORT_SET_POSITION
-        )
+        supported_features = SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_STOP | SUPPORT_SET_POSITION
```
No need for a variable here, just inline with the return.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -207,87 +416,85 @@ def available(self):
-        return self._position
-
     @property
     def is_opening(self):
```
This method can be simplified to:

```python
return self._last_movement == "Opening"
```

Is `"Opening"` something that can vary by the way, so it should be a configuration option?

andrewmeepos replied:
I am not sure if "opening" or "closing" vary. We need information from more cover devices.

**I am going to remove is_opening() and is_closing() for now.** These properties are still unfinished in their implementation. I'm going to work on this in a future PR. There are two types of cover devices we have identified:

(1) Devices like mine that report a wide variety of dps and values, such as the current position, and the last movement of the device ("opening" vs "closing"). For these devices, you can know what the last movement was ("opening" vs "closing") by looking at the dps values. However, you can not definitively know by looking at the dps output whether the device is **currently** "opening" or "closing". We will have to hack together a solution. For example, let's imagine a scenario where the current position of the device is 100 (fully closed) and the last movement is ("closing"). Someone then sets the position to 50. That will immediately change last movement from "closing" to "opening". We could then set is_opening to TRUE until such time as current_cover_position == last_position_set, at which point we set is_opening to FASLE. If someone simply calls open_cover or close_cover, then you can set is_opening or is_closing respectively to TRUE--if someone then calls stop, you know when to stop the is_opening or is_closing. If someone does not calls stop, and lets the cover over or close until it is fully open/fully closed, then we can do some sort of check to see when the current_cover_position updates to 0 or 100. And then there's the edge case, which is what happens when someone controls the cover outside of Home Assistant, using the physical buttons on the device or the device's remote control. I need to investigate how controlling the device in those scenarios modifies dps values (if at all).

(2) Devices like @rospogrigio's that do not let you set or get the position. In essence, all these devices let you do is set them to `on/open` or `off/close` or `stop`. These devices also don't report `opening` or `closing`. For these devices, there's an added complexity, which is that if you open/close and let run until fully open/fully closed, then you can't get the current position to know when it's finished opening/closing. One way around this, perhaps, is to add an optional variable to device setup for number of seconds to fully open/fully close, and we can use that as a max value to determine when the device must have finished opening/closing when a stop was not called.

postlund replied:
This sounds rather complex to implement in a reliable way, especially to handle all odd-cases and variations. Add to that that we have a poll time of one minute (or is it 30s?). So we have a looong feedback loop before we know what is actually happening. As you say, might be good to do that logic in another PR.

andrewmeepos replied:
I'm unsure if it will be reliable, but i'm going to give it a go. In a different PR.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -207,87 +416,85 @@ def available(self):
         return False
 
     @property
     def is_closing(self):
```
See `is_opening`.

andrewmeepos replied:
See my long response to is_opening.

postlund replied:
👍

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -207,87 +416,85 @@ def available(self):
+        _LOGGER.debug("Launching command %s to cover ", self._config[CONF_STOP_CMD])
         self._device.set_dps(self._config[CONF_STOP_CMD], self._dps_id)
 
     def status_updated(self):
```
See comments about `update` earlier. You can remove this and I think you can also remove `self._state` as it doesn't seem to be used anymore.

andrewmeepos replied:
To clarify, what would you like me to remove here?

postlund replied:
You can remove this entire method and rename `update` earlier in the file to `status_updated`.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -207,87 +416,85 @@ def available(self):
         self._state = self.dps(self._dps_id)
+
+
+    # def open_cover(self, **kwargs):
```
Remove old code (everything below as well).

postlund commented on custom_components/localtuya/pytuya/__init__.py:
```diff
@@ -196,6 +196,8 @@ def __init__(self, dev_id, address, local_key, connection_timeout=10):
 
         self.port = 6668  # default - do not expect caller to pass in
+        
+        log.debug("Initiated TuyaDevice where self.id=%s self.address=%s self.local_key=%s self.connection_timeout=%s self.version=%s self.dev_type=%s self.port=%s", self.id, self.address, self.local_key, self.connection_timeout, self.version, self.dev_type, self.port)
```
I would argue the this should be done by the user and not the library (I.e. this log point shouldn't be here).

andrewmeepos replied:
Yes, this was just a debug for me. I'll remove.

postlund replied:
Did you remove this?

postlund commented on custom_components/localtuya/translations/en.json:
```diff
@@ -40,7 +40,10 @@
                     "close_cmd": "Close Command",
-                    "stop_cmd": "Stop Command"
+                    "stop_cmd": "Stop Command",
+                    "get_position_dps": "Get Position DPS",
```
You don't need to include `_DPs` or `DPS` here (update in code as well). It's implicit. It would also have been just `DP` for DataPoint, whereas DPS is DataPoints.

postlund reviewed (changes requested):
Here you have some feedback! 👍

andrewmeepos commented:
@postlund I've incorporated most of your changes. I had a few questions about your feedback, which i've added as comments. I have not yet pushed the updated PR, as I want to resolve my several questions with your feedback before doing so.

postlund commented:
@andrewmeepos I think/hope I answered the questions now.

andrewmeepos commented:
@postlund Ok, I think I updated everything. Take a look and let me know.

andrewmeepos commented:
Would be great to get this merged so I can stop resolving conflicts and/or rebasing over and over 😬

postlund commented on custom_components/localtuya/__init__.py:
```diff
@@ -189,8 +189,10 @@ def update(self):
             self._status = self._device.status()
             self.status_updated()
         except Exception:
+            _LOGGER.debug("update hit exception, setting self._available=false")
```
We need Somme way to de with this, but I don't think this is the right way. It is probably better that we log an exception when things go south but only do it once until the connection is restored. I can sketch something up. Same for the debug below.

andrewmeepos replied:
Ok, i'll just remove these debug lines for now

postlund commented on custom_components/localtuya/config_flow.py:
```diff
@@ -289,8 +287,8 @@ def _convert_entity(conf):
         if len(user_input.get(CONF_SWITCHES, [])) > 0:
-            for switch_conf in user_input[CONF_SWITCHES].values():
-                self.entities.append(_convert_entity(switch_conf))
+                    for switch_conf in user_input[CONF_SWITCHES].values():
```
This indentation doesn't look right. I think you can revert all the changes in this file as nothing should have to change here.

andrewmeepos replied:
Ok, i'll revert config_flow.py to current master

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -32,10 +39,12 @@
 from homeassistant.const import (
     CONF_ID,
     CONF_FRIENDLY_NAME,
+    CONF_NAME,
```
You can remove this, it's not used,

andrewmeepos replied:
Will do

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +201,85 @@ def __init__(
-        return self._available
+    def is_closed(self):
+        """Check if the cover is fully closed."""
+        if self._current_cover_position == 100:
```
Same as:

```python
return self._current_cover_position == 100:
```

andrewmeepos replied:
Good point. I'll clean this up.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +201,85 @@ def __init__(
+        _LOGGER.debug("status_updated called")
+        self._cached_status = self._status
+        _LOGGER.debug("status_updated called, self._status=%s, self._cached_status=%s", self._status, self._cached_status)
+        try:
```
Remember that when you do this (catching exceptions), the cover will still appear as available in the UI and might give the illusion of working, I would argue that it's better that we fail, get an error report and try to solve the problem instead. The logged warning also hides which DP caused the issue.

andrewmeepos replied:
Remember that not all cover devices report the current position, last position set, last command and last movement dp values. I think the best thing right now is to remove the try/except, and then in a future PR, i'll do a specific check to see if the dps exist before trying to update them in status_updated.

postlund replied:
The you should probably use `.get(...)` instead so that `None` is returned when looking up a missing key. Then we at least try to read all values and save whatever we find.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +201,85 @@ def __init__(
-        if state == "on":
-            return True
-        return False
+    def min_position(self):
```
Both `min_position` and `max_position` are not used here and not present in the base class (`CoverEntity`), so you can remove them. The constants can be removed as well.

andrewmeepos replied:
Agreed.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +201,85 @@ def __init__(
-        if state == "on":
-            return True
-        return None
+    def available(self):
```
This is already in `LocalTuyaEntity`, so you can remove it here.

andrewmeepos replied:
👍

postlund commented on custom_components/localtuya/translations/en.json:
```diff
@@ -49,6 +49,9 @@
                     "stop_cmd": "Stop Command",
+                    "get_position": "Get Position",
+                    "set_position": "Set Position",
+                    "last_movement": "Last Movement"
```
You need to add a `,` at the end here.

andrewmeepos replied:
Yup

postlund reviewed (changes requested):
A few small things left but otherwise it's good!

andrewmeepos commented:
@postlund I believe I incorporate all of your comments

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +195,63 @@ def __init__(
-        return supported_features
+    def status_updated(self):
+        """Device status was updated."""
+        self._cached_status = self._status
```
After looking in the file I can see that this isn't necessary anymore and `self._cached_status` can be removed.

postlund commented on custom_components/localtuya/cover.py:
```diff
@@ -180,108 +195,63 @@ def __init__(
+    def status_updated(self):
+        """Device status was updated."""
+        self._cached_status = self._status
+        self._last_movement = self._cached_status['dps'][str(self._config[CONF_LAST_MOVEMENT])]
```
I thought once more about these. You can use `self.dps(...)` to fetch the value (forgot about that one for a second). So basically:

```python
self._last_movement = self.dps(str._config.get(CONF_LAST_MOVEMENT))
```
This will log a warning if the requested DP does not exist.

postlund reviewed (commented):
Added two minor things I found after looking at everybody again. After that I'm all for merging 👍

andrewmeepos commented:
@postlund good to merge?

postlund commented:
@andrewmeepos Yeah! 👍

andrewmeepos commented:
> @andrewmeepos Yeah! 👍

@postlund I tested it and it had some issues, but I just fixed it. Just need to fix conflicts with master, remove my internal logging, and push.

andrewmeepos commented:
@postlund @rospogrigio ok, this should be good to merge

rospogrigio commented:
Sorry @andrewmeepos but I still could not test this and I must ask you to suspend working on this until we have merged #40 because I would really have that sorted before proceeding, and I don't want to waste your time by having you solve conflicts over and over again. Once #40 is done, I promise this will be the next PR to go in. 
Just one comment: I could not test it but I see you have removed the` is_opening()` and `is_closing()` methods, but I quite remember that they need to be defined for my covers to work (I think they are needed for the icons to change when opening/closing). Let me know if your covers behave the same.
Sorry about the mess, please wait #40 before proceeding, thank you.

rospogrigio commented:
Totally rewritten in #60, closing PR.
